contrib/babelfishpg_tds/src/backend/tds/guc.c

```diff
@@ -111,6 +113,62 @@ TdsGucDefaultPacketSizeCheck(int *newvalue, void **extra, GucSource source)
 	return true;
 }
 
+static bool 
+check_version_number(char **newval, void **extra, GucSource source)
+{
+	char 		*copy_version_number = malloc(sizeof(*newval));
+	char		*token;
+	int		part = 0;
+
+    strcpy(copy_version_number,*newval);
+	if(strcasecmp(copy_version_number,"default") == 0)
+		return true;
+
+	for (token = strtok(copy_version_number, "."); token; token = strtok(NULL, "."))
+	{	
+		/* check each token contains only digits */
+		if(strspn(token, "0123456789") != strlen(token))
+		{
+			free(copy_version_number);
+			GUC_check_errmsg("Please enter a valid version number");
+		}
+		
+		/* check Major Version is between 11 and 15 */
+		if(part == 0 && (11 > atoi(token) || atoi(token) > 15))
+		{
+			free(copy_version_number);
+			GUC_check_errmsg("Please enter a valid major version number between 11 and 15");
+		}
+
+		/* Minor Version takes 1 byte in PreLogin message when doing handshake, here to check
+			it is between 0 and 0xFF
+		*/
+		if(part == 1 && (atoi(token) < 0 || atoi(token) > 0xFF))
+		{
+			free(copy_version_number);
+			GUC_check_errmsg("Please enter a valid minor version number");
+		}
+		/* Micro Version takes 2 bytes in PreLogin message when doing handshake, here to check
+			it is between 0 and 0xFFFF
+		*/
+		if(part == 2 && (atoi(token) < 0 || atoi(token) > 0xFFFF))
+		{
+			free(copy_version_number);
+			GUC_check_errmsg("Please enter a valid micro version number");
+		}
+		part++; 
+	}
+
+	if(part != 4)
+	{
+		free(copy_version_number);
+		GUC_check_errmsg("Please enter 4 valid number separated by \'.\' ");
+	}
+
+	free(copy_version_number);
```

Comment:
You can use `PG_TRY/FINALLY` to simplify the code (https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/BABEL_2_X_DEV__PG_14_X/src/include/utils/elog.h#L273-L287).